File: screen_2nd/common/protocol.py
import json
import struct
import zlib
from dataclasses import dataclass, asdict
from typing import Optional, Tuple

UDP_DISCOVERY_PORT = 45678
TCP_DATA_PORT = 45679

MSG_TYPE_HELLO = "hello"
MSG_TYPE_PASSWORD_REQ = "password_req"
MSG_TYPE_PASSWORD_ACK = "password_ack"
MSG_TYPE_DISCONNECT = "disconnect"
MSG_TYPE_FRAME = "frame"
MSG_TYPE_FRAME_ACK = "frame_ack"
MSG_TYPE_TOUCH = "touch"
MSG_TYPE_MODE_CHANGE = "mode_change"

MODE_DISPLAY = "display"
MODE_EXTEND = "extend"
MODE_IDLE = "idle"

PLATFORM_WINDOWS = "windows"
PLATFORM_ANDROID = "android"

DISCOVERY_INTERVAL = 2.0
HEARTBEAT_INTERVAL = 5.0
TIMEOUT = 15.0

COMPRESSION_THRESHOLD = 10240


@dataclass
class DeviceInfo:
    device_id: str
    device_name: str
    platform: str
    ip: str = ""
    tcp_port: int = TCP_DATA_PORT
    has_password: bool = False
    mode: str = MODE_IDLE

    def to_dict(self) -> dict:
        return asdict(self)

    @classmethod
    def from_dict(cls, data: dict) -> "DeviceInfo":
        return cls(
            device_id=data.get("device_id", ""),
            device_name=data.get("device_name", ""),
            platform=data.get("platform", ""),
            ip=data.get("ip", ""),
            tcp_port=data.get("tcp_port", TCP_DATA_PORT),
            has_password=data.get("has_password", False),
            mode=data.get("mode", MODE_IDLE),
        )


@dataclass
class DiscoveryMessage:
    msg_type: str
    device: DeviceInfo

    def to_bytes(self) -> bytes:
        data = {"type": self.msg_type, "device": self.device.to_dict()}
        return json.dumps(data, ensure_ascii=False).encode("utf-8")

    @classmethod
    def from_bytes(cls, data: bytes) -> Optional["DiscoveryMessage"]:
        try:
            obj = json.loads(data.decode("utf-8"))
            device = DeviceInfo.from_dict(obj.get("device", {}))
            return cls(msg_type=obj.get("type", ""), device=device)
        except Exception:
            return None


def pack_message(msg_type: str, payload: bytes = b"") -> bytes:
    header = struct.pack("!II", len(payload), zlib.crc32(payload) & 0xFFFFFFFF)
    type_bytes = msg_type.encode("utf-8")
    type_len = struct.pack("!H", len(type_bytes))
    return header + type_len + type_bytes + payload


def unpack_message(data: bytes) -> Tuple[Optional[str], Optional[bytes], int]:
    if len(data) < 10:
        return None, None, 0
    payload_len, crc = struct.unpack("!II", data[:8])
    type_len = struct.unpack("!H", data[8:10])[0]
    total_len = 10 + type_len + payload_len
    if len(data) < total_len:
        return None, None, 0
    msg_type = data[10 : 10 + type_len].decode("utf-8")
    payload = data[10 + type_len : total_len]
    if zlib.crc32(payload) & 0xFFFFFFFF != crc:
        return None, None, total_len
    return msg_type, payload, total_len


def compress_frame(frame_data: bytes) -> bytes:
    if len(frame_data) > COMPRESSION_THRESHOLD:
        compressed = zlib.compress(frame_data, level=1)
        if len(compressed) < len(frame_data):
            return b"\x01" + compressed
    return b"\x00" + frame_data


def decompress_frame(data: bytes) -> bytes:
    if data[0] == 1:
        return zlib.decompress(data[1:])
    return data[1:]

File: screen_2nd/android_app/app/src/main/java/com/screen2nd/app/network/Protocol.kt
package com.screen2nd.app.network

import org.json.JSONObject
import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.zip.CRC32
import java.util.zip.Deflater
import java.util.zip.Inflater

object Protocol {
    const val UDP_DISCOVERY_PORT = 45678
    const val TCP_DATA_PORT = 45679

    const val MSG_TYPE_HELLO = "hello"
    const val MSG_TYPE_PASSWORD_REQ = "password_req"
    const val MSG_TYPE_PASSWORD_ACK = "password_ack"
    const val MSG_TYPE_DISCONNECT = "disconnect"
    const val MSG_TYPE_FRAME = "frame"
    const val MSG_TYPE_FRAME_ACK = "frame_ack"
    const val MSG_TYPE_TOUCH = "touch"
    const val MSG_TYPE_MODE_CHANGE = "mode_change"

    const val MODE_DISPLAY = "display"
    const val MODE_EXTEND = "extend"
    const val MODE_IDLE = "idle"

    const val PLATFORM_WINDOWS = "windows"
    const val PLATFORM_ANDROID = "android"

    const val DISCOVERY_INTERVAL = 2000L
    const val HEARTBEAT_INTERVAL = 5000L
    const val TIMEOUT = 15000L

    const val COMPRESSION_THRESHOLD = 10240
}

data class DeviceInfo(
    var deviceId: String = "",
    var deviceName: String = "",
    var platform: String = "",
    var ip: String = "",
    var tcpPort: Int = Protocol.TCP_DATA_PORT,
    var hasPassword: Boolean = false,
    var mode: String = Protocol.MODE_IDLE
) {
    fun toJson(): JSONObject {
        return JSONObject().apply {
            put("device_id", deviceId)
            put("device_name", deviceName)
            put("platform", platform)
            put("ip", ip)
            put("tcp_port", tcpPort)
            put("has_password", hasPassword)
            put("mode", mode)
        }
    }

    companion object {
        fun fromJson(json: JSONObject): DeviceInfo {
            return DeviceInfo(
                deviceId = json.optString("device_id", ""),
                deviceName = json.optString("device_name", ""),
                platform = json.optString("platform", ""),
                ip = json.optString("ip", ""),
                tcpPort = json.optInt("tcp_port", Protocol.TCP_DATA_PORT),
                hasPassword = json.optBoolean("has_password", false),
                mode = json.optString("mode", Protocol.MODE_IDLE)
            )
        }
    }
}

data class DiscoveryMessage(
    var msgType: String = "",
    var device: DeviceInfo = DeviceInfo()
) {
    fun toBytes(): ByteArray {
        val json = JSONObject().apply {
            put("type", msgType)
            put("device", device.toJson())
        }
        return json.toString().toByteArray(Charsets.UTF_8)
    }

    companion object {
        fun fromBytes(data: ByteArray): DiscoveryMessage? {
            return try {
                val json = JSONObject(String(data, Charsets.UTF_8))
                DiscoveryMessage(
                    msgType = json.optString("type", ""),
                    device = DeviceInfo.fromJson(json.optJSONObject("device") ?: JSONObject())
                )
            } catch (e: Exception) {
                null
            }
        }
    }
}

object MessagePacker {
    fun pack(msgType: String, payload: ByteArray = byteArrayOf()): ByteArray {
        val typeBytes = msgType.toByteArray(Charsets.UTF_8)
        val crc = CRC32().apply { update(payload) }.value.toInt()

        val buffer = ByteBuffer.allocate(10 + typeBytes.size + payload.size)
        buffer.order(ByteOrder.BIG_ENDIAN)
        buffer.putInt(payload.size)
        buffer.putInt(crc)
        buffer.putShort(typeBytes.size.toShort())
        buffer.put(typeBytes)
        buffer.put(payload)

        return buffer.array()
    }

    fun unpack(data: ByteArray): Triple<String?, ByteArray?, Int> {
        if (data.size < 10) return Triple(null, null, 0)

        val buffer = ByteBuffer.wrap(data)
        buffer.order(ByteOrder.BIG_ENDIAN)

        val payloadLen = buffer.int
        val crc = buffer.int
        val typeLen = buffer.short.toInt()

        val totalLen = 10 + typeLen + payloadLen
        if (data.size < totalLen) return Triple(null, null, 0)

        val typeBytes = ByteArray(typeLen)
        buffer.get(typeBytes)
        val msgType = String(typeBytes, Charsets.UTF_8)

        val payload = ByteArray(payloadLen)
        buffer.get(payload)

        val calcCrc = CRC32().apply { update(payload) }.value.toInt()
        if (calcCrc != crc) {
            return Triple(null, null, totalLen)
        }

        return Triple(msgType, payload, totalLen)
    }
}

object FrameCompressor {
    fun compress(frameData: ByteArray): ByteArray {
        if (frameData.size > Protocol.COMPRESSION_THRESHOLD) {
            val deflater = Deflater(Deflater.BEST_SPEED)
            deflater.setInput(frameData)
            deflater.finish()

            val outputStream = ByteArrayOutputStream(frameData.size)
            val buffer = ByteArray(1024 * 32)

            while (!deflater.finished()) {
                val count = deflater.deflate(buffer)
                outputStream.write(buffer, 0, count)
            }

            deflater.end()
            val compressed = outputStream.toByteArray()

            if (compressed.size < frameData.size) {
                return byteArrayOf(1) + compressed
            }
        }
        return byteArrayOf(0) + frameData
    }

    fun decompress(data: ByteArray): ByteArray {
        if (data[0] == 1.toByte()) {
            val inflater = Inflater()
            inflater.setInput(data, 1, data.size - 1)

            val outputStream = ByteArrayOutputStream()
            val buffer = ByteArray(1024 * 32)

            while (!inflater.finished()) {
                val count = inflater.inflate(buffer)
                outputStream.write(buffer, 0, count)
            }

            inflater.end()
            return outputStream.toByteArray()
        }
        return data.copyOfRange(1, data.size)
    }
}
